sw/applications/i2s/main.c

```diff
@@ -0,0 +1,246 @@
+    CSR_SET_BITS(CSR_REG_MSTATUS, 0x8);
+    // Set mie.MEIE bit to one to enable machine-level external interrupts
+    // bit 11 = external intr (used for i2s)
+    // bit 19 = dma fast intr
+    const uint32_t mask = 1 << 11 | 1 << 19; 
+    CSR_SET_BITS(CSR_REG_MIE, mask);
+}
+
+#define I2S_WAIT_TIME_US 100000 
+#define I2S_WAIT_CYCLES  ((1.5 * REFERENCE_CLOCK_Hz / 4))
+// 1500000
+
+int main(int argc, char *argv[]) {
+#ifdef TARGET_PYNQ_Z2
+    for (uint32_t i = 0; i < 0x10000; i++) asm volatile("nop");
+#endif
+    printf("I2s DEMO\r\n");
+
+    setup();
+
+    //printf("Setup done!\r\n");
+
+#ifdef TARGET_PYNQ_Z2
+    
+
+    for (uint32_t i = 0; i < I2S_WAIT_CYCLES; i++) asm volatile("nop");
+
+    //
+    // FPGA code
+    //
+    #pragma message ( "this application never ends" )
+
+    int batch = 0;
+    while(1) {
+        i2s_rx_start(I2S_LEFT_CH);
+
+        #ifdef USE_DMA
+        dma_set_cnt_start(&dma, (uint32_t) (AUDIO_DATA_NUM*4)); // start 
+
+        // WAITING FOR DMA COPY TO FINISH
+        while(!dma_intr_flag) {
+            wait_for_interrupt();
+        }
+        dma_intr_flag = 0;
+        #else
+        // READING DATA MANUALLY OVER BUS
+        for (int i = 0; i < AUDIO_DATA_NUM; i+=1) {
+            if (i != i2s_rx_read_waterlevel()) printf("Waterlevel wrong\r\n");
+            while (! i2s_rx_data_available()) { }
+            audio_data_0[i] = i2s_rx_read_data();
+        }
+        #endif
+
+
+
+        printf("index,data\r\n"); // <- csv header for python 
+        int32_t* data = audio_data_0;
+        for (int i = 0; i < AUDIO_DATA_NUM; i+=1) {
+            printf("%4x,%d\r\n", i, (int16_t) (data[i] >> 16));
+        }
+        batch += 1;
+
+        if (i2s_interrupt_flag) {
+            printf("irq 1\n");
+            i2s_interrupt_flag = 0;
+        }
+
+        break;
+
+        i2s_rx_stop();
+
+        #ifdef USE_DMA
+        dma_set_cnt_start(&dma, (uint32_t) (AUDIO_DATA_NUM*4)); // restart 
+        #endif
+    }
+#else
+    //
+    // Verilator Code
+    //
+
+    for (int batch = 0; batch < I2S_TEST_BATCHES; batch++) {
+        i2s_rx_start(I2S_BOTH_CH);
+
+        #ifdef USE_DMA
+        dma_set_cnt_start(&dma, (uint32_t) (AUDIO_DATA_NUM*4)); // start 
+
+        // WAITING FOR DMA COPY TO FINISH
+        while(!dma_intr_flag) {
+            wait_for_interrupt();
+        }
+        dma_intr_flag = 0;
+        #else
+        // READING DATA MANUALLY OVER BUS
+        for (int i = 0; i < AUDIO_DATA_NUM; i+=1) {
+            if (i != i2s_rx_read_waterlevel()) printf("Waterlevel wrong\r\n");
+            while (!i2s_rx_data_available()) { }
+            audio_data_0[i] = i2s_rx_read_data();
+        }
+        #endif
+        i2s_rx_stop();
+
+        if (i2s_interrupt_flag) {
+            printf("irq 1\r\n");
+            i2s_interrupt_flag = 0;
+        }
+
+
+        printf("B%x\r\n", batch);
+        
+        int32_t* data = audio_data_0;
+        for (int i = 0; i < AUDIO_DATA_NUM; i+=2) {
```

Comment:
this application always return SUCCESS, so we do not know whether it failed or not, we should rather compare the data got from the tb microphone with expected ones, and if they are wrong, return FAILURE

Reply:
I added two constant values to verify the logic. Plus, a warning if verilator has been built without `--flag=use_external_device_example`